src/pool/srv_pool.c

```diff
@@ -3003,6 +3003,82 @@ ds_pool_svc_get_prop(uuid_t pool_uuid, d_rank_list_t *ranks,
 	return rc;
 }
 
+int
+ds_pool_reintegrate(uuid_t pool_uuid, d_rank_list_t *ranks,
+		uint32_t reint_rank, struct pool_target_id_list *reint_list)
+{
+	int				rc;
+	struct rsvc_client		client;
+	crt_endpoint_t			ep;
+	struct dss_module_info		*info = dss_get_module_info();
+	crt_rpc_t			*rpc;
+	struct pool_target_addr_list	list;
+	struct pool_add_in		*in;
+	struct pool_add_out		*out;
+	int i = 0;
+
+	rc = rsvc_client_init(&client, ranks);
+	if (rc != 0)
+		return rc;
+
+rechoose:
+
+	ep.ep_grp = NULL; /* primary group */
+	rsvc_client_choose(&client, &ep);
+
+	rc = pool_req_create(info->dmi_ctx, &ep, POOL_ADD, &rpc);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to create pool add rpc: "
+			""DF_RC"\n", DP_UUID(pool_uuid), DP_RC(rc));
+		D_GOTO(out_client, rc);
+	}
+
+	in = crt_req_get(rpc);
+	uuid_copy(in->pti_op.pi_uuid, pool_uuid);
+
+	rc = pool_target_addr_list_alloc(reint_list->pti_number, &list);
+	if (rc) {
+		D_ERROR(DF_UUID": pool_target_addr_list_alloc failed, rc %d.\n",
+			DP_UUID(pool_uuid), rc);
+		crt_req_decref(rpc);
+		D_GOTO(out_client, rc);
+	}
+
+	/* pool_update rpc requires an addr list. */
+	for (i = 0; i < reint_list->pti_number; i++) {
+		list.pta_addrs[i].pta_target = reint_list->pti_ids[i].pti_id;
+		list.pta_addrs[i].pta_rank = reint_rank;
+	}
+
+	in->pti_addr_list.ca_arrays = list.pta_addrs;
+	in->pti_addr_list.ca_count = (size_t)list.pta_number;
+
+	rc = dss_rpc_send(rpc);
+	out = crt_reply_get(rpc);
+	D_ASSERT(out != NULL);
+
+	rc = rsvc_client_complete_rpc(&client, &ep, rc,
+		out->pto_op.po_rc, &out->pto_op.po_hint);
+	if (rc == RSVC_CLIENT_RECHOOSE) {
+		crt_req_decref(rpc);
+		dss_sleep(1000 /* ms */);
+		D_GOTO(rechoose, rc);
+	}
+
+	rc = out->pto_op.po_rc;
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to set rank targets up: "DF_RC"\n",
```

Comment:
Another case where it seems like using the "reintegrate" terminology might make the error clearer